leader/leader.go

```diff
@@ -175,51 +185,51 @@ func (l *Client) AddWatch(key string, retry time.Duration, valuesC chan string)
 				}
 			}
 
-			if err != nil {
-				select {
-				case <-ticker.C:
-					steps += 1
+			// The watcher has received a new event, send it to the client
+			// unless this value has already been sent.
+			if err == nil {
+				if resp.PrevNode != nil && resp.PrevNode.Value == resp.Node.Value && sentAnything {
+					continue
 				}
-
-				if err == context.Canceled {
+				select {
+				case valuesC <- resp.Node.Value:
+					sentAnything = true
+				case <-l.closeC:
+					logger.Info("Watcher is closing.")
 					return
-				} else if cerr, ok := err.(*client.ClusterError); ok {
-					if len(cerr.Errors) != 0 && cerr.Errors[0] == context.Canceled {
-						return
-					}
-					log.Debugf("unexpected cluster error: %v (%v)", err, cerr.Detail())
-					continue
-				} else if IsWatchExpired(err) {
-					log.Debug("watch expired, resetting watch index")
-					watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-					if err != nil {
-						continue
-					}
-				} else {
-					log.Warningf("unexpected watch error: %v", err)
-					// try recreating the watch if we get repeated unknown errors
-					if steps > 10 {
-						watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-						if err != nil {
-							continue
-						}
-						backoff.Reset()
-						steps = 0
-					}
-
-					continue
 				}
-			}
-			// if nothing has changed and we previously sent this subscriber this value,
-			// do not bother subscriber with extra notifications
-			if resp.PrevNode != nil && resp.PrevNode.Value == resp.Node.Value && sentAnything {
 				continue
 			}
+
+			// Otherwise, the watcher encountered an error while waiting for
+			// next event, we may need to recreate it.
 			select {
-			case valuesC <- resp.Node.Value:
-				sentAnything = true
-			case <-l.closeC:
+			case <-ticker.C:
```

Comment:
this can degenerate to:
```go
<-ticker.C
steps += 1
```
Are you sure you want to avoid selecting on `l.closeC` here?